plugins/liquid/refined-types-plugin/src/main/kotlin/arrow/meta/plugins/liquid/phases/analysis/solver/check/Expressions.kt

```diff
@@ -932,6 +951,53 @@ private fun SolverState.checkSimpleConditional(
         }
       }
   }
+}
+
+private fun SolverState.introduceCondition(
+  conditionVar: ObjectFormula,
+  subjectVar: ObjectFormula,
+  cond: Condition,
+  data: CheckData
+): ContSeq<Return> = when (cond) {
+  is SimpleCondition -> checkExpressionConstraints(conditionVar, cond.predicate, data)
+  is SubjectCondition -> when (val check = cond.check) {
+    is WhenConditionWithExpression ->
+      if (check.expression is NullExpression) {
+        cont {
+          val complete = solver.booleans {
+            equivalence(solver.boolValue(conditionVar), solver.isNull(subjectVar))
+          }
+          addConstraint(NamedConstraint("$subjectVar is null", complete))
+          NoReturn
+        }
+      } else {
+        val patternName = newName(data.context, "pattern", check.expression)
```

Comment:
Probably easier to understand if we extract big branches of code as separate private named functions